include/deal.II/fe/fe_poly.templates.h

```diff
@@ -569,82 +562,62 @@ FE_Poly<PolynomialType, dim, spacedim>::correct_third_derivatives(
-  {
-    Assert(false, ExcNotImplemented());
-    return std::vector<unsigned int>();
-  }
-
-  template <class PolynomialType>
-  inline std::vector<unsigned int>
-  get_poly_space_numbering_inverse(const PolynomialType &)
-  {
-    Assert(false, ExcNotImplemented());
-    return std::vector<unsigned int>();
-  }
-
-  template <int dim, typename PolynomialType>
-  inline std::vector<unsigned int>
-  get_poly_space_numbering(
-    const TensorProductPolynomials<dim, PolynomialType> &poly)
-  {
-    return poly.get_numbering();
-  }
-
-  template <int dim, typename PolynomialType>
-  inline std::vector<unsigned int>
-  get_poly_space_numbering_inverse(
-    const TensorProductPolynomials<dim, PolynomialType> &poly)
-  {
-    return poly.get_numbering_inverse();
-  }
-
-  template <int dim>
-  inline std::vector<unsigned int>
-  get_poly_space_numbering(const TensorProductPolynomialsConst<dim> &poly)
-  {
-    return poly.get_numbering();
-  }
-
-  template <int dim>
-  inline std::vector<unsigned int>
-  get_poly_space_numbering_inverse(
-    const TensorProductPolynomialsConst<dim> &poly)
-  {
-    return poly.get_numbering_inverse();
-  }
-} // namespace internal
-
-
-
-template <class PolynomialType, int dim, int spacedim>
+
+
+template <int dim, int spacedim>
 std::vector<unsigned int>
-FE_Poly<PolynomialType, dim, spacedim>::get_poly_space_numbering() const
+FE_Poly<dim, spacedim>::get_poly_space_numbering() const
 {
-  return internal::get_poly_space_numbering(poly_space);
+  auto *const space_tensor_prod =
+    dynamic_cast<TensorProductPolynomials<dim, ScalarPolynomialsBase<dim>> *>(
+      this->poly_space.get());
+
+  if (space_tensor_prod != nullptr)
+    return space_tensor_prod->get_numbering();
+
+  auto *const space_tensor_prod_const =
+    dynamic_cast<TensorProductPolynomials<dim, ScalarPolynomialsBase<dim>> *>(
+      this->poly_space.get());
+
+  if (space_tensor_prod_const != nullptr)
+    return space_tensor_prod_const->get_numbering();
+
+  Assert(false, ExcNotImplemented());
+  return std::vector<unsigned int>();
 }
 
 
 
-template <class PolynomialType, int dim, int spacedim>
+template <int dim, int spacedim>
 std::vector<unsigned int>
-FE_Poly<PolynomialType, dim, spacedim>::get_poly_space_numbering_inverse() const
+FE_Poly<dim, spacedim>::get_poly_space_numbering_inverse() const
 {
-  return internal::get_poly_space_numbering_inverse(poly_space);
+  auto *const space_tensor_prod =
+    dynamic_cast<TensorProductPolynomials<dim, ScalarPolynomialsBase<dim>> *>(
+      this->poly_space.get());
+
+  if (space_tensor_prod != nullptr)
+    return space_tensor_prod->get_numbering_inverse();
+
+  auto *const space_tensor_prod_const =
+    dynamic_cast<TensorProductPolynomials<dim, ScalarPolynomialsBase<dim>> *>(
+      this->poly_space.get());
+
+  if (space_tensor_prod_const != nullptr)
+    return space_tensor_prod_const->get_numbering_inverse();
+
+  Assert(false, ExcNotImplemented());
+  return std::vector<unsigned int>();
```

Comment:
Would it make sense to simply return `Utilities::invert_permutation(get_poly_space_numbering())` here?

Reply:
I'm not familiar `invert_permutation`, but it seems like we could use that to replace every instance of `get_numbering_inverse()`, right? This might be a good thing to save for a follow-up PR.

Reply:
Sorry I misread this. I just realized this function is what broke the majority of the testsuite, although I'm unsure why. I pushed a change with your suggestion. It builds fine on my home machine so I hope this reduces the trouble in the testsuite. :)

Reply:
It looks like this change causes all of the errors to propagate to `get_poly_space_numbering()` instead unfortunately. I'll have to look at the casts more carefully and see why it's failing.